tests/src/Clash/Vivado.hs

```diff
@@ -0,0 +1,149 @@
+{-# LANGUAGE QuasiQuotes #-}
+
+-- | Generate a TCL script to simulate generated VHDL
+--
+-- Run with @vivado -mode batch -source ...@
+module Clash.Vivado ( HdlSource (..), fromModuleName ) where
+
+import Paths_clash_testsuite (getDataDir)
+
+import Clash.Driver.Manifest (readManifest, Manifest (..))
+import Clash.Annotations.Primitive (HDL (..))
+
+import Data.String.Interpolate (i)
+import Data.List (isSuffixOf)
+
+import System.FilePath ((</>))
+import qualified Data.Text as T
+
+data HdlSource = HdlSource { directory :: FilePath, buildTarget :: HDL }
+
+data SimProj = SimProj { vhdlFiles :: [(T.Text, FilePath)]
+                       , verilogFiles :: [(T.Text, FilePath)]
+                       , ipGen :: [FilePath]
+                       } deriving Show
+
+instance Semigroup SimProj where
+  (<>) (SimProj vhd0 v0 ip0) (SimProj vhd1 v1 ip1) =
+    SimProj (vhd0<>vhd1) (v0<>v1) (ip0<>ip1)
+
+instance Monoid SimProj where
+  mempty = SimProj [] [] []
+
+thread :: [a -> a] -> a -> a
+thread = foldr (.) id
+
+-- | Given something like @topEntity_dcfifo_240A2A9D3FF8D64B.tcl@, return the
+-- IP name, @dcfifo@
+parseTcl :: FilePath -> String
+parseTcl fp = takeWhile (/= '_') (drop 1 $ dropWhile (/= '_') fp)
+
+mSimProj :: FilePath -> T.Text -> FilePath -> SimProj -> SimProj
+mSimProj dir lib fp p@(SimProj vhd v ip)
+  | ".vhdl" `isSuffixOf` fp = SimProj ((lib, dir </> fp):vhd) v ip
+  | ".vhd" `isSuffixOf` fp = SimProj ((lib, dir </> fp):vhd) v ip
+  | ".v" `isSuffixOf` fp = SimProj vhd ((lib, dir </> fp):v) ip
+  | ".tcl" `isSuffixOf` fp = SimProj vhd v ((dir </> fp):ip)
+  | otherwise = p
+
+simProjFromClashEntities ::
+  -- | HDL source dir
+  FilePath ->
+  -- | Qualified name
+  String ->
+  IO SimProj
+simProjFromClashEntities hdlDir qualName = do
+  (Just mHdl) <- readManifest (hdlDir </> qualName ++ "/clash-manifest.json")
+  let deps = T.unpack <$> transitiveDependencies mHdl
+  nextSimProj <- traverse (simProjFromClashEntities hdlDir) deps
+  pure $ mconcat nextSimProj <> asSimProj qualName mHdl
+
+fromModuleName ::
```

Comment:
I think this function has an unclear name and there's no documentation either. `mkTcl` feels appropriate for this and then rename the existing `mkTcl`? I dunno.

Reply:
I've named it `tclFromManifest` which I think makes sense.